Sources/Amplitude/Amplitude.m

```diff
@@ -465,33 +465,39 @@ - (void)initializeApiKey:(NSString *)apiKey
             }
         }];
 
-        // Normally _inForeground is set by the enterForeground callback, but initializeWithApiKey will be called after the app's enterForeground
-        // notification is already triggered, so we need to manually check and set it now.
-        // UIApplication methods are only allowed on the main thread so need to dispatch this synchronously to the main thread.
-        void (^checkInForeground)(void) = ^{
-        #if !TARGET_OS_OSX && !TARGET_OS_WATCH
-            UIApplication *app = [AMPUtils getSharedApplication];
-            if (app != nil) {
-                UIApplicationState state = app.applicationState;
-                if (state != UIApplicationStateBackground) {
-                    [self runOnBackgroundQueue:^{
-        #endif
-                        // The earliest time to fetch dynamic config
-                        [self refreshDynamicConfig];
-                        
-                        NSNumber *now = [NSNumber numberWithLongLong:[[self currentTime] timeIntervalSince1970] * 1000];
-                        [self startOrContinueSessionNSNumber:now];
-                        self->_inForeground = YES;
-        #if !TARGET_OS_OSX && !TARGET_OS_WATCH
-                    }];
+        if (!self.deferCheckInForeground) {
+            [self checkInForeground];
+        }
+    }
+}
+
+- (void) checkInForeground {
+    // Normally _inForeground is set by the enterForeground callback, but initializeWithApiKey will be called after the app's enterForeground
+    // notification is already triggered, so we need to manually check and set it now.
+    // UIApplication methods are only allowed on the main thread so need to dispatch this synchronously to the main thread.
+    void (^checkInForeground)(void) = ^{
+    #if !TARGET_OS_OSX && !TARGET_OS_WATCH
+        UIApplication *app = [AMPUtils getSharedApplication];
+        if (app != nil) {
+            UIApplicationState state = app.applicationState;
+            if (state != UIApplicationStateBackground) {
+                [self runOnBackgroundQueue:^{
+    #endif
+                    // The earliest time to fetch dynamic config
+                    [self refreshDynamicConfig];
+
+                    NSNumber *now = [NSNumber numberWithLongLong:[[self currentTime] timeIntervalSince1970] * 1000];
+                    [self startOrContinueSessionNSNumber:now];
+                    self->_inForeground = YES;
+    #if !TARGET_OS_OSX && !TARGET_OS_WATCH
+                }];
 
-                }
             }
-        #endif
-        };
-        [self runSynchronouslyOnMainQueue:checkInForeground];
-        _initialized = YES;
-    }
+        }
+    #endif
+    };
+    [self runSynchronouslyOnMainQueue:checkInForeground];
+    _initialized = YES;
```

Comment:
_initialized is only used in [here](https://github.com/amplitude/Amplitude-iOS/blob/main/Sources/Amplitude/Amplitude.m#L448) to skip the checkInForeground logic and set identity logic if already initialized. 
Since the checkInForeground is a part of initialization and other logic is related to set deviceId and userId. I think it will be fine. But definitely want to hear @justin-fiedler 's opinion.

Reply:
Thanks @yuhao900914 I think your changes look good, especially since it is likely only going to be used by a single customer for now.

Can you please add a section to the docs on usage? Or maybe confirm the solution works for the customer first, then add docs

Reply:
Sure, will publish this first and confirm with the customer. I will add the usage in the developer center after.